qubes.lua.in

```diff
@@ -170,60 +161,61 @@ function qubes.get_colour_focus(c)
     return get_colour_by_state(c, 'qubes_label_color_focus')
 end
 
-function qubes.manager_menuitem()
-    return {'Qubes Manager', 'qubes-manager',
-        '/usr/share/icons/hicolor/16x16/apps/qubes-logo-icon.png'}
+function is_accessible(path)
+    return os.rename(path, path) and true or false
 end
 
 function qubes.make_menuitem(vm, program)
-    return {program.Name, program.cmdline, program.icon_path}
+    local appicon = program.icon_path
+    if not is_accessible(appicon) then appicon = nil end
+
+    return {program.Name, program.cmdline, appicon}
 end
 
-function qubes.make_vm_menu(vm, menuitemfunc)
+--all parameters except for vmname are optional
+function qubes.make_vm_menu(vmname, menuitemfunc, vmapps, vmicon)
     local menu = {}
+    local vmapps = vmapps or HOME .. '/.local/share/qubes-appmenus/' .. vmname .. '/apps/'
```

Comment:
Why don't use files installed in standard location: `~/.local/share/applications`? Or more generally - why this code is needed at all - does Awesome not support standard menu from `.desktop` and `.directory` and `.menu` files (i.e. XDG standard)?

Using standard location will work on any Qubes version.

Reply:
Honestly I just continued to do what was previously done in this repo. I think the approach was implemented as awesome supports .desktop files, but no .menu or .directory files.
@woju Can you confirm my assumption?

In 3.2 `/var/lib/qubes/[vm]/apps` was parsed by the `qubes.lua` code. The VM names were used as the first level of menu items. In 4.0 that structure was apparently moved to `~/.local/share/qubes-appmenus/`, so I used that one.

So yes, we can just use `~/.local/share/applications`, but that would then create menu entries which are all on the first level and create a rather messy menu as all imposed hierarchy from the .menu files is lost.

It doesn't look like there are plans by awesome to support .menu files any time soon - at least I didn't find any related issue on their bugtracker. By default they seem to support categories for some menu structure [1]. That's also what other menu implementations seem to use ( e.g. [2] [3]), but isn't really helpful for a VM-based hierarchy.

Some proper .menu file parsing would of course be the best solution, but require quite a lot of work I guess.

If you don't want to stick to the old approach, I'd recommend to just reject this PR.

[1] https://awesomewm.org/apidoc/libraries/menubar.menu_gen.html
[2] https://wiki.archlinux.org/index.php/Xdg-menu
[3] https://github.com/lcpz/awesome-freedesktop

Reply:
I'm trying to avoid the need to port this code to every new Qubes version. Using `.menu` files should accomplish it, even if just minimal support is here. I don't know lua (and its support for XML), but for shell something like this would work:

For each menu file:
1. Get menu name:
```
xmllint --xpath '/Menu/Menu/Name/text()' /path/to/file.menu
```
2. Get list of included desktop files:
```
echo '/Menu/Menu/Include/Filename/text()' | xmllint --shell /path/to/file.menu | grep 'desktop$'
```

Then, look for those files in `~/.local/share/applications/`. Rest of the code should be the same.

Reply:
So essentially you suggest to parse the `.menu` files at `~/.config/menus/applications-merged` and use the directory mentioned in them as 1st level menu, then parse the `.desktop` files at `~/.local/share/applications` and use the entries there as 2nd level menu?

Maybe that's easily feasible with some lua regex, I'll check.

Reply:
> So essentially you suggest to parse the .menu files at ~/.config/menus/applications-merged and use the directory mentioned in them as 1st level menu, then parse the .desktop files at ~/.local/share/applications and use the entries there as 2nd level menu?

Yes, exactly.

Reply:
I changed that now, but didn't test too much yet.